File: 20190917/20190920/119. 杨辉三角 II.py
'''
Description: 杨辉三角 II
Author: AskeyNil
CreateDate: 2019-09-20 08:19:59
LastEditors: AskeyNil

*********************************
**                             **
**     　  你只有足够努力     　  **
**       才能看上去毫不费力       **
**                             **
*********************************

'''


# 递归法
# class Solution:
#     def getRow(self, rowIndex: int) -> [int]:
#         if rowIndex == 0:
#             return [1]
#         else:
#                 row.append(num + left)
#             left = 0
#             rows = self.getRow(rowIndex - 1)
#             row = []
#             for num in rows:
#                 left = num
#             return row
#             row.append(1)


# 公式法
class Solution:
    def getRow(self, rowIndex: int) -> [int]:
        rows = [1]
        if rowIndex == 1:
            return rows
        left_n = 1
        left_k = 1
        for index in range(rowIndex):
            left_n *= (rowIndex - index)
            left_k *= index + 1
            rows.append(left_n // left_k)
        return rows


print(Solution().getRow(5))

File: 20190917/20190920/119. 杨辉三角 II.cc
/*
 * Description: 杨辉三角 II
 * Author: AskeyNil
 * CreateDate: 2019-09-20 08:27:33
 * LastEditors: AskeyNil
 *
 * *********************************
 * **                             **
 * **     　  你只有足够努力     　  **
 * **       才能看上去毫不费力       **
 * **                             **
 * *********************************
 *
 */

#include <iostream>
#include <string>
#include <vector>

using namespace std;

// 递归法
// class Solution {
//    public:
//     vector<int> getRow(int rowIndex) {
//         if (rowIndex == 0) {
//             vector<int> vv;
//             vv.push_back(1);
//             return vv;
//         } else {
//             vector<int> vv = this->getRow(rowIndex - 1);
//             vector<int> vv1;
//             int left = 0;
//             for (auto num : vv) {
//                 vv1.push_back(left + num);
//                 left = num;
//             }
//             vv1.push_back(1);
//             return vv1;
//         }
//     }
// };

// 公式法
class Solution {
   public:
    vector<int> getRow(int rowIndex) {
        vector<int> v{1};
        if (rowIndex == 0) return v;
        for (int i = 0; i < rowIndex; i++) {
            if (i >= rowIndex / 2.0) {
                v.push_back(v[rowIndex - i - 1]);
            } else {
                v.push_back(v[i] * ((long)rowIndex - i) / (i + 1));
            }
        }
        return v;
    }
};

ostream& operator<<(ostream& out, vector<int> v) {
    for (auto i : v) {
        out << i << endl;
    }
    return out;
}

int main(int argc, char const* argv[]) {
    std::cout << Solution().getRow(100) << std::endl;
    return 0;
}
